swift-sdk/Internal/IterableAPIInternal.swift

```diff
@@ -399,56 +399,34 @@ final class IterableAPIInternal : NSObject, PushTrackerProtocol {
         }
     }
     
-    func spawn(inAppNotification callbackBlock:ITEActionBlock?) {
-        let onSuccess: OnSuccessHandler = { payload in
-            guard let payload = payload else {
-                return
-            }
-            guard let dialogOptions = IterableInAppManager.getNextMessageFromPayload(payload) else {
-                ITBError("No notifications found for inApp payload \(payload)")
-                return
-            }
-            guard let message = dialogOptions[.ITBL_IN_APP_CONTENT] as? [AnyHashable : Any] else {
-                return
-            }
-            guard let messageId = dialogOptions[.ITBL_KEY_MESSAGE_ID] as? String else {
-                return
-            }
-            guard let html = message[.ITBL_IN_APP_HTML] as? String else {
-                return
-            }
-            guard html.range(of: AnyHashable.ITBL_IN_APP_HREF, options: [.caseInsensitive]) != nil else {
-                ITBError("No href tag found in in-app html payload \(html)")
-                self.inAppConsume(messageId)
-                return
-            }
-
-            let inAppDisplaySettings = message[.ITBL_IN_APP_DISPLAY_SETTINGS] as? [AnyHashable : Any]
-            let backgroundAlpha = IterableInAppManager.getBackgroundAlpha(fromInAppSettings: inAppDisplaySettings)
-            let edgeInsets = IterableInAppManager.getPaddingFromPayload(inAppDisplaySettings)
-
-            let notificationMetadata = IterableNotificationMetadata.metadata(fromInAppOptions: messageId)
-            
-            DispatchQueue.main.async {
-                let opened = IterableInAppManager.showIterableNotificationHTML(html, trackParams: notificationMetadata, backgroundAlpha: backgroundAlpha, padding: edgeInsets, callbackBlock: callbackBlock)
-                if opened {
-                    self.inAppConsume(messageId)
+    @discardableResult func spawn(inAppNotification callbackBlock:ITEActionBlock?) -> Future<IterableInAppManager.ShowInAppResult> {
+        return getInAppMessages(1)
+            .map { IterableInAppManager.parseInApp(fromPayload: $0)}
+            .flatMap { IterableInAppManager.showInApp(parseResult: $0, callbackBlock: callbackBlock)}
+            .onSuccess {
+                switch $0 {
+                case .success(opened: let opened, messageId: let messageId):
+                    if opened {
+                        self.inAppConsume(messageId)
+                    }
+                case .failure(reason: let reason, messageId: let messageId):
+                    if let messageId = messageId {
+                        self.inAppConsume(messageId)
```

Comment:
Hmm looks like it wasn't consuming the in-app if it was not shown, but now it'll consume it even when it fails

Reply:
Oh, it looks like it will return success with opened = false if it was shown, but failure if it could not be parsed.. Was it doing the same before the change?

Reply:
Actually it will return success with opened = true if it was shown, false otherwise (if there is already an in-app message showing etc.).
If there is an error, it will return with messageId if one is available. 
This was the behavior before the change.

Reply:
Correct that was the original behavior. If the content of the html is invalid we don't attempt to show it and we remove it from the queue. IE: if the html doesn't have a href button then don't show it otherwise there is no way to close the dialog.